Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+        if (numHandlers < 1)
+        {
+            return;
+        }
+
+        ComponentData newComponentData = ComponentData(pairId);
+
+        // If the component has not already been added as a button, add the button
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [newComponentData](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_component->RTTI_GetType() == newComponentData.m_component->RTTI_GetType();
+                });
+            componentDataIt == m_addedComponents.end())
+        {
+            AddSwitcherButton(newComponentData, newComponentData.m_componentName.c_str(), newComponentData.m_iconPath.c_str());
+            m_addedComponents.push_back(newComponentData);
+        }
+    }
+
+    void ComponentModeSwitcher::AddSwitcherButton(ComponentData& componentData, const char* componentName, const char* iconPath)
+    {
+        ViewportUi::ButtonId buttonId;
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            buttonId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcherButton,
+            m_switcherId,
+            iconPath,
+            componentName);
+
+        componentData.m_buttonId = buttonId;
+
+        AZStd::string buttonTooltip = AZStd::string::format(" Enter %s component mode", componentData.m_componentName.c_str());
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherButtonTooltip,
+            m_switcherId,
+            buttonId,
+            buttonTooltip);
+    }
+
+    void ComponentModeSwitcher::RemoveComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // find pairId in m_whatever, call bus to remove
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [pairId](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_pairId == pairId;
+                });
+            componentDataIt != m_addedComponents.end())
+        {
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::RemoveSwitcherButton,
+                m_switcherId,
+                componentDataIt->m_buttonId);
+
+            m_addedComponents.erase(componentDataIt);
+        }
+    }
+
+    void ComponentModeSwitcher::ActivateComponentMode(const ViewportUi::ButtonId buttonId)
+    {
+        bool inComponentMode;
+        AzToolsFramework::ComponentModeFramework::ComponentModeSystemRequestBus::BroadcastResult(
+            inComponentMode, &ComponentModeSystemRequests::InComponentMode);
```

Comment:
suggestion: Always good to keep variables close to first use, so we might as well move this block down to be just above line 239 as that's where we use this value